pyccel/parser/semantic.py

```diff
@@ -2746,12 +2850,47 @@ def _visit_While(self, expr, **settings):
         return While(test, body, local_vars)
 
     def _visit_IfSection(self, expr, **settings):
-        cond = self._visit(expr.condition)
+        condition = expr.condition
+
+        name_symbol = PyccelSymbol('__name__')
+        main = LiteralString('__main__')
+        prog_check = isinstance(condition, PyccelEq) \
+                and all(a in (name_symbol, main) for a in condition.args)
+
+        if prog_check:
```

Comment:
can you check this test
```python
def func(x:int):
    if __name__ == '__main__':
        print('this is stupid thing')
    return x * x
```

Reply:
I get:
```
ERROR at code generation stage
pyccel:
 |fatal [codegen]: main_odd.py| _print_InProgram is not yet implemented for language : Fortran
Pyccel has encountered syntax that has not been implemented yet. Please create an issue at https://github.com/pyccel/pyccel/issues and provide a small example of your problem. Do not forget to specify your target language (<pyccel.ast.core.InProgram object at 0x7fd07b8934c0>
```
So it is correctly creating this object in the function rather than the module and is then not unpacking the contents into the module. The statement `if __name__ == '__main__':` then raises an error when we try to print it.

I think this is acceptable behaviour for something that a user is very unlikely to do, but if you prefer I could add `_print_InProgram` methods to raise a more specific error?

Reply:
while python letting that for the user, we may do too.
the function can have different behavior if it's imported or called inside the main block.
but as you said this is rare, so I think it's better to raise an error and let the user try another good way instead of this stupid method xD.